docs/type_hints.rst

```diff
@@ -0,0 +1,100 @@
+
+For clarify, the following is an example of a function ``foo`` with type hints:
+
+.. code-block:: python
+	
+    def foo(a: int) -> int:
+        return a + 1
+
+While the following is an example of a type alias ``Bar``:
+
+.. code-block:: python
+
+    from typing import Tuple
+
+    Bar = Tuple[int, str]
+
+RDFLib will provide public type aliases under the ``rdflib.typing`` package, for example, ``rdflib.typing.Triple``, ``rdflib.typing.Quad``. Type aliases in the rest of RDFLib should be private (i.e. being with an underscore).
+
+Versioning, Compatibility and Stability
+=======================================
+
+RDFLib attempts to adhere to `semver 2.0 <https://semver.org/spec/v2.0.0.html>`_ which is concerned with the public API of software.
+
+Ignoring type hints, the public API of RDFLib exists implicitly as a consequence of the code of RDFLib and the actual behaviour this entails, the relevant standards that RDFLib is trying to implement, and the documentation of RDFLib, with some interplay between all three of these. RDFLib's public API includes public type aliases, as these are normal python variables and not annotations.
+
+Type hints attempt to formally document RDFLib's implicitly defined public API in a machine-readable fashion as accurately and correctly as possible within the framework outline earlier in this document.
+
+Type hints do not change the functionality of RDFLib, nor do they affect the runtime API of RDFLib. In this way then, they are somewhat outside of the scope of semver, however, they still have an impact on the users of RDFLib, even if this impact is not at runtime, but during development. This necessitates some clarity as to what users of RDFLib should expect regarding type hints in RDFLib releases.
+
+Changes to type hints can broadly be classified as follow:
+
+* **Type Addition**: Adding new type hints to existing code.
+* **Type Narrowing**: Changing type hints of existing code to be narrower. The line between type additions and narrowing is somewhat blurry, as un-annotated code is usually treated the same as if it was annotated with `typing.Any`, so type addition is a form of type narrowing.
+* **Type Corrections** to type hints which contradict the behavior of the code or relevant specifications.
+
+Given semver version components ``MAJOR.MINOR.PATCH``, RDFLib will attempt to constrain type hint changes as follow:
+
+.. list-table::
+   :widths: 1 1 1 1
+   :header-rows: 1
+
+   * - Version Component
+     - Type Additions
+     - Type Narrowing
+     - Type Corrections
+
+   * - MAJOR
+     - YES
+     - YES
+     - YES
+  
+   * - MINOR
+     - YES
+     - YES
+     - YES
+
+   * - PATCH
+     - NO
+     - NO
+     - YES
+
+A caveat worth nothing here is that code that passed type validation on one version of RDFLib can fail type validation on a later version of RDFLib that only differs in ``PATCH`` version component.
```

Comment:
I'm pondering this caveat still, but I'm currently thinking written here might be the best that can be done for downstream projects.  I'm personally fine with this being the merge-able wording.

Reply:
Hah, and going back to look at my prior comment, I see I didn't change my mind.

Reply:
...I might be missing something, but I don't see the button to close this thread.  I think it can be closed.